### DISCUSSION TITLE
git init and git add .  Inadvertently added submodule. How can I correct this?

### DISCUSSION THREAD

**Discussion by dmastrop** [community/community#51876] [Programming Help]
### Body

hi all.
I am going through some docker training and in this multi-container project there is a main folder /complex and then several subfolders for the respective containers.

See below

This is after doing a git init and then a git add . in the parent directory /complex.

The client sub-directory was created earlier with an npx create-react-app client and I think it created a local git repo at that time. I see a .git folder in the /complex/client directory and can only assume this is what happened.

Here is an ls -la of the /complex root directory of this project after running the 
git init
git add .

As you can see it added its own git repo.  But the git add . threw an error (see next output below) 

complex % ls -la
total 24

 **.git**
 .github
**client**
docker-compose.yml
nginx
server
t.travis.yml_Section10_travis_only
test
worker

=======
Here is the error. It ended up creating a submodule as the /complex/client already had its own local git repo

this git add . was done from the /complex root directory

git add .
**warning: adding embedded git repository: client**
hint: You've added another git repository inside your current repository.
hint: Clones of the outer repository will not contain the contents of
hint: the embedded repository and will not know how to obtain it.
hint: If you meant to add a submodule, use:
hint:
hint:   git submodule add <url> client
hint:
hint: I**f you added this path by mistake, you can remove it from the
hint: index with:
hint:
hint:   git rm --cached client**
hint:
hint: See "git help submodule" for more information.

======
here is the current git status.  You can clearly see that the contents of the /client directory are untracked.
I have not made any commits yet at all to the repo. So it is still pretty clean but before proceeding with any fix I wanted to make sure the commands will work.  Is this recoverable assuming no commits have been done yet.

On branch master

**No commits yet**

Changes to be committed:
  (use "git rm --cached <file>..." to unstage)
        **new file:   client**  <<< I think just the directory itself but not the contents, see below for contents (not tracked)
        new file:   docker-compose.yml
        new file:   nginx/Dockerfile
        new file:   nginx/Dockerfile.dev
        new file:   nginx/default.conf
        new file:   server/Dockerfile
        new file:   server/Dockerfile.dev
        new file:   server/index.js
        new file:   server/keys.js
        new file:   server/package.json
        new file:   t.travis.yml_Section10_travis_only
        new file:   test
        new file:   worker/Dockerfile
        new file:   worker/Dockerfile.dev
        new file:   worker/index.js
        new file:   worker/keys.js
        new file:   worker/package.json

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        **modified:   client (modified content, untracked content)**  <<<<<<<<<<<<<<<<<<<<<<<<<<
        modified:   t.travis.yml_Section10_travis_only

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        .github/

=========

I did some research. would this resolve this issue?. I just want one local git repo for the entire project /complex and no submodules, etc.
The other subfolders are ok (no submodule git repos on those, as there is no .git in their folders)

Executing these from the /complex parent directory

**git rm --client
git commit -m "remove cached client repo"    <<<<< will this delete the entire directory or just remove it from the parent repo?
git add client/    <<<< note must add the trailing /  I read that the trailing / is the key.
git commit -m "Add client folder"**

Assuming the other pending staged files above are now committed:

Do a **git status** to verify and then 

**Git add .github/
Git add t.travis.yml_Section10_travis_only**

**git commit -m "commit the rest of the stuff"**

=========

Here is the current /complex/client directory

 client % ls -la
total 1368
.dockerignore
**.git**
.gitignore
Dockerfile
Dockerfile.dev
README.md
nginx
node_modules
package-lock.json
package.json
public
src

===========

_Also, what is the correct way to do this from the onset to avoid getting into this situation. Can the .git file be manually deleted from the subdirectory /complex/client before doing the git init and git add . in the /complex parent directory?_

warm regards

### I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

- [X] Yes

**Comment by airtower-luna** [comment]
> Also, what is the correct way to do this from the onset to avoid getting into this situation.

Don't use `git add .`. The dot in that command means "the current directory" and will add everything inside it. It's extremely easy to make mistakes that way. Instead `git add` the files you explicitly want to add.

> Can the .git file be manually deleted from the subdirectory /complex/client before doing the git init and git add . in the /complex parent directory?

If you never meant `/complex/client` to be a Git repository of its own: Yes. Keep in mind though that `.git` is what holds all the repository history, so only delete it if you're absolutely sure you won't need it again. :warning: 

As long as you haven't committed anything, you can just use `git reset` to undo the `git add`. Clean up the Git repository inside `/complex/client` if you want, and then add the files you want to add. :slightly_smiling_face:

**Comment by sergeyleschev** [comment]
You can use `git rm --cached <submodule-name>` to remove it. If you haven't committed anything yet, you can also use `git reset` to undo the `git add`. After cleaning up the Git repository inside `/complex/client`, you can then add the files you want to add using `git add <directory-name>/` and commit the changes.
